Surface tracker crashes attempt to overwrite cached position

<details>
<summary>Crash log</summary>
```
2020-02-06 12:05:34,297 - player - [ERROR] launchables.player: Process Player crashed with trace:
Traceback (most recent call last):
  File "launchables/player.py", line 608, in player
  File "shared_modules/surface_tracker/surface_tracker_offline.py", line 234, in recent_events
  File "shared_modules/surface_tracker/surface_tracker.py", line 430, in recent_events
  File "shared_modules/surface_tracker/surface_tracker_offline.py", line 263, in _update_markers
  File "shared_modules/surface_tracker/surface_tracker_offline.py", line 285, in _update_marker_and_surface_caches
  File "shared_modules/surface_tracker/cache.py", line 48, in update
IndexError: Can not overwrite an already cached position without force!
```
</details>

--- a/pupil_src/shared_modules/surface_tracker/offline_utils.py
+++ b/pupil_src/shared_modules/surface_tracker/offline_utils.py
@@ -14,6 +14,18 @@ from .surface_marker_detector import MarkerDetectorController
 
 
 class marker_detection_callable(MarkerDetectorController):
+    @classmethod
+    def from_detector(cls, detector: MarkerDetectorController, marker_min_perimeter):
+        return cls(
+            marker_detector_mode=detector.marker_detector_mode,
+            marker_min_perimeter=marker_min_perimeter,
+            square_marker_inverted_markers=detector.inverted_markers,
+            square_marker_use_online_mode=False,
+            apriltag_nthreads=detector._apriltag_nthreads,
+            apriltag_quad_decimate=detector.apriltag_quad_decimate,
+            apriltag_decode_sharpening=detector.apriltag_decode_sharpening,
+        )
+
     def __call__(self, frame):
         return self.detect_markers(gray_img=frame.gray, frame_index=frame.index)
 

--- a/pupil_src/shared_modules/surface_tracker/surface_marker_detector.py
+++ b/pupil_src/shared_modules/surface_tracker/surface_marker_detector.py
@@ -13,7 +13,6 @@ import abc
 import enum
 import logging
 import typing as T
-from collections import namedtuple
 
 import square_marker_detect
 import pupil_apriltags
@@ -47,11 +46,10 @@ class ApriltagFamily(enum.Enum):
     tagStandard52h13 = "tagStandard52h13"
 
 
-# TODO: Use T.NamedTuple as soon as we increase Python version requirement to 3.6.1+
-_MarkerDetectorMode = namedtuple("_MarkerDetectorMode", ["marker_type", "family"])
+class MarkerDetectorMode(T.NamedTuple):
+    marker_type: Surface_Marker_Type
+    family: T.Optional[str]
 
-
-class MarkerDetectorMode(_MarkerDetectorMode):
     @classmethod
     def all_supported_cases(cls) -> T.Set["MarkerDetectorMode"]:
         all_square = {cls(MarkerType.SQUARE_MARKER, None)}

--- a/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
+++ b/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
@@ -15,6 +15,7 @@ import multiprocessing
 import os
 import platform
 import time
+import typing as T
 
 import cv2
 import numpy as np
@@ -49,6 +50,12 @@ if platform.system() == "Darwin":
     mp_context = multiprocessing.get_context("fork")
 else:
     mp_context = multiprocessing.get_context()
+
+
+class _CacheRelevantDetectorParams(T.NamedTuple):
+    inverted_markers: bool
+    quad_decimate: float
+    sharpening: float
 
 
 class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):

--- a/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
+++ b/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
@@ -173,9 +173,12 @@ class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):
             logger.debug("Restored previous marker cache.")
 
     def _set_detector_params(self, params: _CacheRelevantDetectorParams):
-        self.inverted_markers = params.inverted_markers
-        self.quad_decimate = params.quad_decimate
-        self.sharpening = params.sharpening
+        self.marker_detector._marker_detector_mode = params.mode
+        self.marker_detector._square_marker_inverted_markers = params.inverted_markers
+        self.marker_detector._apriltag_quad_decimate = params.quad_decimate
+        self.marker_detector._apriltag_decode_sharpening = params.sharpening
+        self.marker_detector.init_detector()
+
     @staticmethod
     def _cache_relevant_params_from_cache(
         previous_cache,
@@ -613,7 +616,9 @@ class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):
             self.marker_cache_unfiltered
         )
         marker_cache_file["version"] = self.MARKER_CACHE_VERSION
-        marker_cache_file["inverted_markers"] = self.inverted_markers
-        marker_cache_file["quad_decimate"] = self.quad_decimate
-        marker_cache_file["sharpening"] = self.sharpening
+
+        current_config = self._cache_relevant_params_from_controller()
+        marker_cache_file["inverted_markers"] = current_config.inverted_markers
+        marker_cache_file["quad_decimate"] = current_config.quad_decimate
+        marker_cache_file["sharpening"] = current_config.sharpening
         marker_cache_file.save()

--- a/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
+++ b/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
@@ -82,8 +82,6 @@ class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):
         self.last_cache_update_ts = time.perf_counter()
         self.CACHE_UPDATE_INTERVAL_SEC = 5
 
-        self._init_marker_detection_modes()
-
         self.gaze_on_surf_buffer = None
         self.gaze_on_surf_buffer_filler = None
 
@@ -122,24 +120,26 @@ class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):
     def supported_heatmap_modes(self):
         return [Heatmap_Mode.WITHIN_SURFACE, Heatmap_Mode.ACROSS_SURFACES]
 
-    def _init_marker_detection_modes(self):
-        # This method should be called after `_init_marker_cache` to ensure that the cache is filled in.
-        assert self.marker_cache is not None
+    @staticmethod
+    def _marker_detector_mode_from_cache(
+        marker_cache,
+    ) -> T.Optional[MarkerDetectorMode]:
+        assert marker_cache is not None
 
         # Filter out non-filled frames where the cache entry is None.
         # Chain the remaining entries (which are lists) to get a flat sequence.
-        filled_out_marker_cache = filter(lambda x: x is not None, self.marker_cache)
+        filled_out_marker_cache = filter(lambda x: x is not None, marker_cache)
         cached_surface_marker_sequence = itertools.chain.from_iterable(
             filled_out_marker_cache
         )
 
-        # Get the first surface marker from the sequence, and set the detection mode according to it.
-        first_cached_surface_marker = next(cached_surface_marker_sequence, None)
-        if first_cached_surface_marker is not None:
-            marker_detector_mode = MarkerDetectorMode.from_marker(
-                first_cached_surface_marker
-            )
-            self.marker_detector.marker_detector_mode = marker_detector_mode
+        # Get the first surface marker from the sequence, and set the detection mode
+        # according to it.
+        first_cached_surface_marker_args = next(cached_surface_marker_sequence, None)
+        if first_cached_surface_marker_args is not None:
+            marker = Surface_Marker.deserialize(first_cached_surface_marker_args)
+            marker_detector_mode = MarkerDetectorMode.from_marker(marker)
+            return marker_detector_mode
 
     def _init_marker_cache(self):
         previous_cache_config = file_methods.Persistent_Dict(
@@ -175,6 +175,25 @@ class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):
         self.inverted_markers = params.inverted_markers
         self.quad_decimate = params.quad_decimate
         self.sharpening = params.sharpening
+    @staticmethod
+    def _cache_relevant_params_from_cache(
+        previous_cache,
+    ) -> T.Optional[_CacheRelevantDetectorParams]:
+        marker_cache_unfiltered = previous_cache.get("marker_cache_unfiltered", None)
+        if marker_cache_unfiltered is None:
+            return
+
+        mode = Surface_Tracker_Offline._marker_detector_mode_from_cache(
+            marker_cache_unfiltered
+        )
+        if mode is None:
+            return
+
+        return _CacheRelevantDetectorParams(
+            inverted_markers=previous_cache.get("inverted_markers", False),
+            quad_decimate=previous_cache.get("quad_decimate", APRILTAG_HIGH_RES_ON),
+            sharpening=previous_cache.get("sharpening", APRILTAG_SHARPENING_ON),
+        )
 
     def _cache_relevant_params_from_controller(self) -> _CacheRelevantDetectorParams:
         return _CacheRelevantDetectorParams(

--- a/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
+++ b/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
@@ -142,21 +142,23 @@ class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):
             self.marker_detector.marker_detector_mode = marker_detector_mode
 
     def _init_marker_cache(self):
-        previous_cache = file_methods.Persistent_Dict(
+        previous_cache_config = file_methods.Persistent_Dict(
             os.path.join(self.g_pool.rec_dir, "square_marker_cache")
         )
-        version = previous_cache.get("version", 0)
-        cache = previous_cache.get("marker_cache_unfiltered", None)
+        version = previous_cache_config.get("version", 0)
         self._set_detector_params_from_previous_cache(previous_cache)
+        cached_markers_unfiltered = previous_cache_config.get(
+            "marker_cache_unfiltered", None
+        )
 
-        if cache is None:
+        if cached_markers_unfiltered is None:
             self._recalculate_marker_cache()
         elif version != self.MARKER_CACHE_VERSION:
             logger.debug("Marker cache version missmatch. Rebuilding marker cache.")
             self._recalculate_marker_cache()
         else:
             marker_cache_unfiltered = []
-            for markers in cache:
+            for markers in cached_markers_unfiltered:
                 # Loaded markers are either False, [] or a list of dictionaries. We
                 # need to convert the dictionaries into Surface_Marker objects.
                 if markers:

--- a/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
+++ b/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
@@ -225,15 +225,11 @@ class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):
 
         if self.cache_filler is not None:
             self.cache_filler.cancel()
+
         self.cache_filler = background_tasks.background_video_processor(
             self.g_pool.capture.source_path,
-            offline_utils.marker_detection_callable(
-                marker_detector_mode=self.marker_detector.marker_detector_mode,
-                marker_min_perimeter=self.CACHE_MIN_MARKER_PERIMETER,
-                square_marker_inverted_markers=self.inverted_markers,
-                square_marker_use_online_mode=False,
-                apriltag_quad_decimate=self.quad_decimate,
-                apriltag_decode_sharpening=self.sharpening,
+            offline_utils.marker_detection_callable.from_detector(
+                self.marker_detector, self.CACHE_MIN_MARKER_PERIMETER
             ),
             list(self.marker_cache),
             self.cache_seek_idx,

--- a/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
+++ b/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
@@ -53,6 +53,7 @@ else:
 
 
 class _CacheRelevantDetectorParams(T.NamedTuple):
+    mode: MarkerDetectorMode
     inverted_markers: bool
     quad_decimate: float
     sharpening: float
@@ -190,6 +191,7 @@ class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):
             return
 
         return _CacheRelevantDetectorParams(
+            mode=mode,
             inverted_markers=previous_cache.get("inverted_markers", False),
             quad_decimate=previous_cache.get("quad_decimate", APRILTAG_HIGH_RES_ON),
             sharpening=previous_cache.get("sharpening", APRILTAG_SHARPENING_ON),
@@ -197,6 +199,7 @@ class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):
 
     def _cache_relevant_params_from_controller(self) -> _CacheRelevantDetectorParams:
         return _CacheRelevantDetectorParams(
+            mode=self.marker_detector.marker_detector_mode,
             inverted_markers=self.marker_detector.inverted_markers,
             quad_decimate=self.marker_detector.apriltag_quad_decimate,
             sharpening=self.marker_detector.apriltag_decode_sharpening,

--- a/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
+++ b/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
@@ -147,18 +147,18 @@ class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):
             os.path.join(self.g_pool.rec_dir, "square_marker_cache")
         )
         version = previous_cache_config.get("version", 0)
-        cached_markers_unfiltered = previous_cache_config.get(
-            "marker_cache_unfiltered", None
-        )
 
-        if cached_markers_unfiltered is None:
-            self._recalculate_marker_cache()
+        previous_params = self._cache_relevant_params_from_cache(previous_cache_config)
+        current_params = self._cache_relevant_params_from_controller()
+
+        if previous_params is None:
+            self._recalculate_marker_cache(parameters=current_params)
         elif version != self.MARKER_CACHE_VERSION:
             logger.debug("Marker cache version missmatch. Rebuilding marker cache.")
-            self._recalculate_marker_cache()
+            self._recalculate_marker_cache(parameters=current_params)
         else:
             marker_cache_unfiltered = []
-            for markers in cached_markers_unfiltered:
+            for markers in previous_cache_config["marker_cache_unfiltered"]:
                 # Loaded markers are either False, [] or a list of dictionaries. We
                 # need to convert the dictionaries into Surface_Marker objects.
                 if markers:
@@ -169,7 +169,9 @@ class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):
                     ]
                 marker_cache_unfiltered.append(markers)
 
-            self._recalculate_marker_cache(previous_state=marker_cache_unfiltered)
+            self._recalculate_marker_cache(
+                parameters=previous_params, previous_state=marker_cache_unfiltered
+            )
             logger.debug("Restored previous marker cache.")
 
     def _set_detector_params(self, params: _CacheRelevantDetectorParams):
@@ -208,8 +210,11 @@ class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):
             sharpening=self.marker_detector.apriltag_decode_sharpening,
         )
 
-    def _recalculate_marker_cache(self, previous_state=None):
-        self._set_detector_params(self._cache_relevant_params_from_controller())
+    def _recalculate_marker_cache(
+        self, parameters: _CacheRelevantDetectorParams, previous_state=None
+    ):
+        # Ensures consistency across foreground and background detectors
+        self._set_detector_params(parameters)
         if previous_state is None:
             previous_state = [None for _ in self.g_pool.timestamps]
 
@@ -519,7 +524,8 @@ class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):
         super().on_notify(notification)
 
         if notification["subject"] == "surface_tracker.marker_detection_params_changed":
-            self._recalculate_marker_cache()
+            current_params = self._cache_relevant_params_from_controller()
+            self._recalculate_marker_cache(parameters=current_params)
 
         elif notification["subject"] == "surface_tracker.marker_min_perimeter_changed":
             marker_type = self.marker_detector.marker_detector_mode.marker_type

--- a/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
+++ b/pupil_src/shared_modules/surface_tracker/surface_tracker_offline.py
@@ -146,7 +146,6 @@ class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):
             os.path.join(self.g_pool.rec_dir, "square_marker_cache")
         )
         version = previous_cache_config.get("version", 0)
-        self._set_detector_params_from_previous_cache(previous_cache)
         cached_markers_unfiltered = previous_cache_config.get(
             "marker_cache_unfiltered", None
         )
@@ -172,12 +171,20 @@ class Surface_Tracker_Offline(Observable, Surface_Tracker, Plugin):
             self._recalculate_marker_cache(previous_state=marker_cache_unfiltered)
             logger.debug("Restored previous marker cache.")
 
-    def _set_detector_params_from_previous_cache(self, previous_cache):
-        self.inverted_markers = previous_cache.get("inverted_markers", False)
-        self.quad_decimate = previous_cache.get("quad_decimate", APRILTAG_HIGH_RES_ON)
-        self.sharpening = previous_cache.get("sharpening", APRILTAG_SHARPENING_ON)
+    def _set_detector_params(self, params: _CacheRelevantDetectorParams):
+        self.inverted_markers = params.inverted_markers
+        self.quad_decimate = params.quad_decimate
+        self.sharpening = params.sharpening
+
+    def _cache_relevant_params_from_controller(self) -> _CacheRelevantDetectorParams:
+        return _CacheRelevantDetectorParams(
+            inverted_markers=self.marker_detector.inverted_markers,
+            quad_decimate=self.marker_detector.apriltag_quad_decimate,
+            sharpening=self.marker_detector.apriltag_decode_sharpening,
+        )
 
     def _recalculate_marker_cache(self, previous_state=None):
+        self._set_detector_params(self._cache_relevant_params_from_controller())
         if previous_state is None:
             previous_state = [None for _ in self.g_pool.timestamps]
 

